### PR TITLE
Add unit tests for Fisher bias and Δν vector (#138)

### PR DESCRIPTION
**Pull request by nikosarcevic**
This update adds a full set of tests for the Fisher-bias calculation and the Δν (data-difference) helper. The tests cover both simple and edge cases to ensure stable behavior under different covariance and model setups.

What’s included:

- Identity covariance case that reproduces ordinary least squares results
- Non-identity (weighted) covariance case matching generalized least squares
- Consistency check for 2D Δν flattening
- Singular Fisher and singular covariance fallbacks using pseudoinverse
- Input shape validation and handling of non-finite data

All tests passed locally (109 passed, 1 skipped)
Ruff checks passed cleanly

this builds on top of PR #127 and solves #138

### CODE REVIEW

File: tests/test_expansions.py [hunk 1]
```diff
@@ -357,3 +358,198 @@ def model(theta):  # returns length 3
...
+
+    le2 = LikelihoodExpansion(model_2, theta0=np.zeros(2), cov=cov_2)
+
+    data_with = np.array([3.0, -1.0], dtype=float)
+    data_without = np.array([2.5, -2.0], dtype=float)
+    delta_1d = le2.build_delta_nu(data_with, data_without)
+    np.testing.assert_allclose(delta_1d, np.array([0.5, 1.0], dtype=float))
+    assert delta_1d.shape == (le2.n_observables,)
+
+    # 2D case → flattened length 6 (n_observables = 6)
```
Comment (lonbar) [inline tests/test_expansions.py]: Please write in using grammatical english using full stentences.

File: tests/test_expansions.py [hunk 2]
```diff
@@ -357,3 +358,198 @@ def model(theta):  # returns length 3
...
+    return design_matrix @ theta
+
+
+def test_build_delta_nu_1d_and_2d_row_major():
+    """Test that build_delta_nu returns correct shapes and values."""
+    # 1D case (n_observables = 2)
+    cov_2 = np.eye(2)
+
+    def model_2(theta):
+        return np.asarray([theta[0], theta[1]], dtype=float)
```
Comment (lonbar) [inline tests/test_expansions.py]: This is a function of class 1.

File: tests/test_expansions.py [hunk 3]
```diff
@@ -357,3 +358,198 @@ def model(theta):  # returns length 3
...
+
+def test_build_delta_nu_validation_errors():
+    """Test build_delta_nu raises on bad inputs."""
+    cov = np.eye(2)
+    def linear_model(theta):
+        return np.asarray([theta[0], theta[1]], dtype=float)
+
+    le = LikelihoodExpansion(linear_model, theta0=np.zeros(2), cov=cov)
+
+    with pytest.raises(ValueError):
```
Comment (lonbar) [inline tests/test_expansions.py]:
```suggestion
    with pytest.raises(ValueError):
    # This should raise an exception because the two data vectors are of incompatible length
```

File: tests/test_expansions.py [hunk 4]
```diff
@@ -357,3 +358,198 @@ def model(theta):  # returns length 3
...
+    data_without = np.array([2.5, -2.0], dtype=float)
+    delta_1d = le2.build_delta_nu(data_with, data_without)
+    np.testing.assert_allclose(delta_1d, np.array([0.5, 1.0], dtype=float))
+    assert delta_1d.shape == (le2.n_observables,)
+
+    # 2D case → flattened length 6 (n_observables = 6)
+    cov_6 = np.eye(6)
+
+    def model_6(theta):
+        return np.zeros(6, dtype=float)  # model output not used by build_delta_nu
```
[the remark this answers is not in the document]
Comment (lonbar) [reply tests/test_expansions.py]: The comment confuses me a little; if the model output is not used here, why is it necessary in `test_build_delta_nu_validation_errors`?

File: tests/test_expansions.py [hunk 5]
```diff
@@ -357,3 +358,198 @@ def model(theta):  # returns length 3
...
+
+    # Ground truth for identity covariance would be ordinary least squares:
+    expected_bias = design_matrix.T @ delta_nu
+    theta_lstsq, *_ = np.linalg.lstsq(design_matrix, delta_nu, rcond=None)
+
+    np.testing.assert_allclose(bias_vec, expected_bias, rtol=1e-12, atol=1e-12)
+    np.testing.assert_allclose(delta_theta, theta_lstsq, rtol=1e-10, atol=1e-12)
+
+
+def test_fisher_bias_matches_gls_weighted_cov():
```
Comment (lonbar) [inline tests/test_expansions.py]: The same comments at `test_fisher_bias_matches_lstsq_identity_cov` apply here. Perhaps both functions can be merged?

File: tests/test_expansions.py [hunk 6]
```diff
@@ -357,3 +358,198 @@ def model(theta):  # returns length 3
...
+
+    np.testing.assert_allclose(bias_vec, expected_bias, rtol=1e-12, atol=1e-12)
+    np.testing.assert_allclose(delta_theta, expected_delta_theta, rtol=1e-10, atol=1e-12)
+
+
+def test_fisher_bias_accepts_2d_delta_row_major_consistency():
+    """Test that build_fisher_bias accepts 2D delta_nu and matches flattened 1D input."""
+    design_matrix = np.array([[1.0, 2.0],
+                              [0.5, -1.0]], float)
+    def model(theta): return design_matrix @ np.asarray(theta, float)
```
Comment (lonbar) [inline tests/test_expansions.py]: This function is identical to `_linear_model` so is unnecessary.

**Comment by nikosarcevic** [comment]
> I think these tests are pretty comprehensive, but the way the model functions are constructed could be tidied up. Would there be a strong reason to not use `_linear_model` for all models? I think it can be used for all cases with suitable choices of `design_matrix`.

done in bulk in 10b46413504f88dab923320bc4c93f95ff3ddf0c

File: tests/test_expansions.py [hunk 7]
```diff
@@ -357,3 +358,198 @@ def model(theta):  # returns length 3
...
+    """Test that build_fisher_bias raises on mismatched shapes."""
+    def model(theta): return np.asarray([theta[0], theta[1]], float)
+    le = LikelihoodExpansion(model, theta0=np.zeros(2), cov=np.eye(2))
+    fisher = np.eye(3)
+    with pytest.raises(ValueError):
+        le.build_fisher_bias(fisher_matrix=fisher, delta_nu=np.zeros(2))
+
+    fisher_ok = np.eye(2)
+    with pytest.raises(ValueError):
+        le.build_fisher_bias(fisher_matrix=fisher_ok, delta_nu=np.zeros(3))
```
[the remark this answers is not in the document]
Comment (nikosarcevic) [reply tests/test_expansions.py]: fixed in ec3effcf5bd11d82b6ae59d6ba466f80dc9798ec

File: tests/test_expansions.py [hunk 8]
```diff
@@ -357,3 +358,198 @@ def model(theta):  # returns length 3
...
+    covariance = np.array(
+        [[2.0, 0.3],
+         [0.3, 1.0]],
+        dtype=float,
+    )
+    inv_covariance = np.linalg.inv(covariance)
+
+    le = LikelihoodExpansion(linear_model, theta0=np.zeros(3), cov=covariance)
+
+    # Fisher matrix computed by the class with the stored covariance.
```
[the remark this answers is not in the document]
Comment (nikosarcevic) [reply tests/test_expansions.py]: outdated, removed with 10b46413504f88dab923320bc4c93f95ff3ddf0c
